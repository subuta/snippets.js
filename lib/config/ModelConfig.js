import check from 'lib/_utils/apiCheck'

const ModelConfigShape = check.shape({
  schema: check.object
})

// eg:
// config = {
//   book: {
//     schema: {}
//   }
// }
export const validateModels = function (config) {
  return check.throw(check.objectOf(ModelConfigShape), config)
}

// eg:
// config = {
//   schema: {}
// }
export const validateModel = function (config) {
  return check.throw(ModelConfigShape, config)
}
