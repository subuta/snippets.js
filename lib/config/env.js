import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'

export default (imports = []) => {
  const vars = _.zipObject(imports, imports)

  return build`
    // load .env first.
    require('dotenv').config()
    
    const ${s.raw(vars)} = process.env
    
    export ${s.raw(vars)}
    
    export default ${s.raw(vars)}
  `
}
