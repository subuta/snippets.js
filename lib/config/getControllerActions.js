import _ from 'lodash'

export default (config = {}) => {
  const {
    only, // the routes will generated.
    except  // the routes will ignored.
  } = config

  const routes = [
    'index',
    'show',
    'create',
    'update',
    'destroy'
  ]

  if (only) {
    return _.intersection(routes, only)
  } else if (except) {
    return _.difference(routes, except)
  }

  return routes
}
