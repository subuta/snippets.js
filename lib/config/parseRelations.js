import _ from 'lodash'
import pluralize from 'pluralize'

export default (config = {}) => {
  const { schema } = config

  const relations = _.transform(schema.relations, (result, value, key) => {
    let relatedModel = null;
    if (value.belongsTo) {
      relatedModel = pluralize.singular(_.toLower(value.belongsTo))
    } else if (value.hasOne) {
      relatedModel = pluralize.singular(_.toLower(value.hasOne))
    } else if (value.hasMany) {
      relatedModel = pluralize(_.toLower(value.hasMany))
    } else if (value.hasAndBelongsToMany) {
      relatedModel = pluralize(_.toLower(value.hasAndBelongsToMany))
    }
    result[key] = relatedModel
  }, {})

  return relations
}
