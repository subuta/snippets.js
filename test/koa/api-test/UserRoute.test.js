/* global expect, describe, it, beforeEach, afterEach */

import { build, format, snippets as s } from 'bld.js'

import UserRouteTest from 'lib/koa/api-test/UserRoute'

import { Routes as RoutesConfig, Models as ModelsConfig } from 'test/fixtures/config'
import sinon from 'sinon'

describe('koa/api/UserRoute', () => {
  let clock

  beforeEach(() => {
    clock = sinon.useFakeTimers()
  })

  afterEach(() => {
    clock.restore()
  })

  it('should create koa UserRoute', () => {
    expect(format(UserRouteTest({
      model: 'user',
      routeConfig: RoutesConfig.user,
      modelConfig: ModelsConfig.user
    }))).toMatchSnapshot()
  })
})
