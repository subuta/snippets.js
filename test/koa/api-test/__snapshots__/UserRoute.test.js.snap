// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`koa/api/UserRoute should create koa UserRoute 1`] = `
"import test from 'ava'
import _ from 'lodash'
import sinon from 'sinon'
import request from 'supertest'
import {jwksEndpoint} from 'jwks-rsa/tests/mocks/jwks'
import {publicKey, privateKey} from 'jwks-rsa/tests/mocks/keys'
import {createToken} from 'jwks-rsa/tests/mocks/tokens'
import Koa from 'koa'
import importFresh from 'import-fresh'
import {absolutePath} from '../../config'
import {currentUser} from 'test/helper/user'
import runSeed, {runMigration} from 'test/helper/fixtures'
import proxyquire from 'proxyquire'

const sandbox = sinon.sandbox.create()

test.beforeEach(async (t) => {
  const knex = importFresh(absolutePath('src/utils/knex')).default

  await runMigration(knex)
  await runSeed(knex)

  const api = require('test/helper/mocked').api(knex)
  const models = require('test/helper/mocked').model(knex)

  const app = new Koa()
  // handle /api requests
  app.use(api.routes())
  app.use(api.allowedMethods())

  t.context = {
    ...models,
    request: request(app.listen(0))
  }
})

test.afterEach((t) => {
  sandbox.reset()
})

test('get me should return user', async (t) => {
  const {request} = t.context

  // mock jwks
  const token = createToken(privateKey, '123', currentUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .get('/api/users/me')
    .set('Authorization', \`Bearer \${token}\`)

  t.is(response.status, 200)
  t.deepEqual(response.body.id, 2054)
})

test('put me should update user if exists', async (t) => {
  const {request, User} = t.context

  const user = await User.query().first({auth0Id: 'google-oauth2|dummy'})

  // mock jwks
  const token = createToken(privateKey, '123', currentUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .put('/api/users/me')
    .set('Authorization', \`Bearer \${token}\`)
    .send({
      user: {
        id: 2054,
        auth0Id: 40550,
        locale: 'zh_TW',
        nickname: 'Puerto Rico Kids',
        status: 'CFA Franc BEAC',
        avatar:
          'https://s3.amazonaws.com/uifaces/faces/twitter/samscouto/128.jpg',
        created_at: '2000-02-02T09:44:39.710Z',
        updated_at: '2000-12-12T14:14:34.366Z'
      }
    })

  t.is(response.status, 200)

  t.deepEqual(response.body.id, 2054)
  t.deepEqual(response.body.auth0Id, 40550)
  t.deepEqual(response.body.locale, 'zh_TW')
  t.deepEqual(response.body.nickname, 'Puerto Rico Kids')
  t.deepEqual(response.body.status, 'CFA Franc BEAC')
  t.deepEqual(
    response.body.avatar,
    'https://s3.amazonaws.com/uifaces/faces/twitter/samscouto/128.jpg'
  )
  t.deepEqual(response.body.created_at, '2000-02-02T09:44:39.710Z')
  t.deepEqual(response.body.updated_at, '2000-12-12T14:14:34.366Z')
})

test('put me should create user if not exists', async (t) => {
  const {request, User} = t.context

  let user = await User.query().findFirst({auth0Id: 'google-oauth2|another'})
  t.is(user, undefined)

  const anotherUser = {
    iss: 'https://xxx.com/',
    sub: 'google-oauth2|another',
    aud: ['https://xxx.com/api', 'https://xxx.auth0.com/userinfo'],
    scope: 'openid profile email'
  }

  // mock jwks
  const token = createToken(privateKey, '123', anotherUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .put('/api/users/me')
    .set('Authorization', \`Bearer \${token}\`)
    .send({
      user: {
        id: 2054,
        auth0Id: 40550,
        locale: 'zh_TW',
        nickname: 'Puerto Rico Kids',
        status: 'CFA Franc BEAC',
        avatar:
          'https://s3.amazonaws.com/uifaces/faces/twitter/samscouto/128.jpg',
        created_at: '2000-02-02T09:44:39.710Z',
        updated_at: '2000-12-12T14:14:34.366Z'
      }
    })

  t.is(response.status, 200)

  t.deepEqual(response.body.id, 2054)
  t.deepEqual(response.body.auth0Id, 40550)
  t.deepEqual(response.body.locale, 'zh_TW')
  t.deepEqual(response.body.nickname, 'Puerto Rico Kids')
  t.deepEqual(response.body.status, 'CFA Franc BEAC')
  t.deepEqual(
    response.body.avatar,
    'https://s3.amazonaws.com/uifaces/faces/twitter/samscouto/128.jpg'
  )
  t.deepEqual(response.body.created_at, '2000-02-02T09:44:39.710Z')
  t.deepEqual(response.body.updated_at, '2000-12-12T14:14:34.366Z')
})
"
`;
